File: packages/shared/src/types/api/transition-pathways.ts
/**
 * PathForge — API Types: Transition Pathways
 * ============================================
 * Types for career transitions, skill bridges, milestones, and comparisons.
 * Mirrors: apps/api/app/schemas/transition_pathways.py
 */

// ── Skill Bridge ───────────────────────────────────────────────

export interface SkillBridgeEntryResponse {
  readonly id: string;
  readonly skill_name: string;
  readonly category: string;
  readonly is_already_held: boolean;
  readonly current_level: string | null;
  readonly required_level: string | null;
  readonly acquisition_method: string | null;
  readonly estimated_weeks: number | null;
  readonly recommended_resources: Record<string, unknown> | null;
  readonly priority: string;
  readonly impact_on_confidence: number | null;
}

// ── Milestones ─────────────────────────────────────────────────

export interface TransitionMilestoneResponse {
  readonly id: string;
  readonly phase: string;
  readonly title: string;
  readonly description: string | null;
  readonly target_week: number;
  readonly order_index: number;
  readonly is_completed: boolean;
  readonly completed_at: string | null;
}

// ── Comparison ─────────────────────────────────────────────────

export interface TransitionComparisonResponse {
  readonly id: string;
  readonly dimension: string;
  readonly source_value: number;
  readonly target_value: number;
  readonly delta: number;
  readonly unit: string | null;
  readonly reasoning: string | null;
}

// ── Transition Path ────────────────────────────────────────────

export interface TransitionPathResponse {
  readonly id: string;
  readonly career_dna_id: string;
  readonly from_role: string;
  readonly to_role: string;
  readonly confidence_score: number;
  readonly difficulty: string;
  readonly status: string;
  readonly skill_overlap_percent: number;
  readonly skills_to_acquire_count: number;
  readonly estimated_duration_months: number | null;
  readonly optimistic_months: number | null;
  readonly realistic_months: number | null;
  readonly conservative_months: number | null;
  readonly salary_impact_percent: number | null;
  readonly success_probability: number;
  readonly reasoning: string | null;
  readonly factors: Record<string, unknown> | null;
  readonly data_source: string;
  readonly disclaimer: string;
  readonly computed_at: string;
}

// ── Summary (for lists/dashboard) ──────────────────────────────

export interface TransitionSummaryResponse {
  readonly id: string;
  readonly from_role: string;
  readonly to_role: string;
  readonly confidence_score: number;
  readonly difficulty: string;
  readonly status: string;
  readonly skill_overlap_percent: number;
  readonly estimated_duration_months: number | null;
  readonly computed_at: string;
}

// ── Preferences ────────────────────────────────────────────────

export interface TransitionPreferenceResponse {
  readonly id: string;
  readonly preferred_industries: string[] | null;
  readonly excluded_roles: string[] | null;
  readonly min_confidence: number;
  readonly max_timeline_months: number;
  readonly notification_enabled: boolean;
}

export interface TransitionPreferenceUpdateRequest {
  readonly preferred_industries?: string[] | null;
  readonly excluded_roles?: string[] | null;
  readonly min_confidence?: number;
  readonly max_timeline_months?: number;
  readonly notification_enabled?: boolean;
}

// ── Request Schemas ────────────────────────────────────────────

export interface TransitionExploreRequest {
  readonly target_role: string;
  readonly target_industry?: string | null;
  readonly target_location?: string | null;
}

export interface TransitionCompareRequest {
  readonly target_roles: string[];
}

export interface RoleWhatIfRequest {
  readonly target_role: string;
}

// ── Composite Responses ────────────────────────────────────────

export interface TransitionScanResponse {
  readonly transition_path: TransitionPathResponse;
  readonly skill_bridge: SkillBridgeEntryResponse[];
  readonly milestones: TransitionMilestoneResponse[];
  readonly comparisons: TransitionComparisonResponse[];
}

export interface TransitionDashboardResponse {
  readonly transitions: TransitionSummaryResponse[];
  readonly preferences: TransitionPreferenceResponse | null;
  readonly total_explored: number;
}

File: apps/api/app/schemas/transition_pathways.py
"""
PathForge — Transition Pathways Schemas
==========================================
Pydantic schemas for the Transition Pathways API.

Request/response validation with strict typing, transparency
fields (data_source, disclaimer), and confidence clamping.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# ── Response Schemas ───────────────────────────────────────────


class SkillBridgeEntryResponse(BaseModel):
    """Individual skill gap entry in the Skill Bridge Matrix™."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    skill_name: str
    category: str
    is_already_held: bool
    current_level: str | None = None
    required_level: str | None = None
    acquisition_method: str | None = None
    estimated_weeks: int | None = None
    recommended_resources: dict[str, Any] | None = None
    priority: str
    impact_on_confidence: float | None = None


class TransitionMilestoneResponse(BaseModel):
    """Action plan milestone in the Transition Timeline Engine™."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    phase: str
    title: str
    description: str | None = None
    target_week: int
    order_index: int
    is_completed: bool
    completed_at: datetime | None = None


class TransitionComparisonResponse(BaseModel):
    """Source vs target role comparison dimension."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    dimension: str
    source_value: float
    target_value: float
    delta: float
    unit: str | None = None
    reasoning: str | None = None


class TransitionPathResponse(BaseModel):
    """Full transition path with Transition Confidence Score™."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    career_dna_id: uuid.UUID
    from_role: str
    to_role: str
    confidence_score: float
    difficulty: str
    status: str
    skill_overlap_percent: float
    skills_to_acquire_count: int
    estimated_duration_months: int | None = None
    optimistic_months: int | None = None
    realistic_months: int | None = None
    conservative_months: int | None = None
    salary_impact_percent: float | None = None
    success_probability: float
    reasoning: str | None = None
    factors: dict[str, Any] | None = None
    data_source: str
    disclaimer: str
    computed_at: datetime


class TransitionSummaryResponse(BaseModel):
    """Lightweight transition card for list views."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    from_role: str
    to_role: str
    confidence_score: float
    difficulty: str
    status: str
    skill_overlap_percent: float
    estimated_duration_months: int | None = None
    computed_at: datetime


class TransitionPreferenceResponse(BaseModel):
    """User transition preference state."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    preferred_industries: list[str] | None = None
    excluded_roles: list[str] | None = None
    min_confidence: float
    max_timeline_months: int
    notification_enabled: bool


# ── Request Schemas ────────────────────────────────────────────


class TransitionExploreRequest(BaseModel):
    """Request to explore a career transition to a target role."""

    target_role: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Target role to explore transition to",
    )
    target_industry: str | None = Field(
        None,
        max_length=255,
        description="Optional target industry for context",
    )
    target_location: str | None = Field(
        None,
        max_length=255,
        description="Optional target location for context",
    )


class TransitionCompareRequest(BaseModel):
    """Request to compare current role against multiple targets."""

    target_roles: list[str] = Field(
        ...,
        min_length=1,
        max_length=5,
        description="Target roles to compare (max 5)",
    )


class RoleWhatIfRequest(BaseModel):
    """Quick exploration: 'What if I moved to role X?'"""

    target_role: str = Field(
        ...,
        min_length=2,
        max_length=255,
        description="Target role for what-if analysis",
    )


class TransitionPreferenceUpdateRequest(BaseModel):
    """Partial update for transition preferences."""

    preferred_industries: list[str] | None = None
    excluded_roles: list[str] | None = None
    min_confidence: float | None = Field(None, ge=0.0, le=1.0)
    max_timeline_months: int | None = Field(None, ge=1, le=120)
    notification_enabled: bool | None = None


# ── Composite Response Schemas ─────────────────────────────────


class TransitionScanResponse(BaseModel):
    """Full pipeline result: path + skills + milestones + comparison."""

    transition_path: TransitionPathResponse
    skill_bridge: list[SkillBridgeEntryResponse]
    milestones: list[TransitionMilestoneResponse]
    comparisons: list[TransitionComparisonResponse]


class TransitionDashboardResponse(BaseModel):
    """All saved transitions + preferences for the dashboard."""

    transitions: list[TransitionSummaryResponse]
    preferences: TransitionPreferenceResponse | None = None
    total_explored: int
